src/main/scala/codechicken/multipart/util/MergedVoxelShapeHolder.java

```diff
@@ -31,16 +31,23 @@ public void clear() {
     }
 
     public VoxelShape update(Collection<T> things, Function<T, VoxelShape> extractor) {
-        partCache.clear();
-        for (T thing : things) {
-            partCache.add(extractor.apply(thing));
-        }
-        if (!partCache.equals(shapeParts) || mergedShape == null) {
-            shapeParts.clear();
-            shapeParts.addAll(partCache);
-            //Same as VoxelShapes.or(VoxelShapes.empty(), shapeParts.toArray()); Except we skip useless array creation.
-            VoxelShape merged = shapeParts.stream().reduce(VoxelShapes.empty(), VoxelShapes::or);
-            mergedShape = postProcess.apply(merged);
+        synchronized (partCache) {
+            partCache.clear();
+            for (T thing : things) {
+                partCache.add(extractor.apply(thing));
+            }
+
+            if (!partCache.equals(shapeParts) || mergedShape == null) {
+                VoxelShape merged;
```

Comment:
Another minor nitpick. (Sorry!)
This should be inlined below to keep the patch a simple wrap in `synchronized`.

Reply:
Ah, yep. Another change I forgot to undo when I expanded the first synchronized scope